File: LeetCode/Interview Questions/C++/RotateArray.cpp
#include <iostream>
#include <vector>

using namespace std;

void reverse(vector<int>&, int, int);

void rotate(vector<int>& nums, int k) {
    k %= nums.size();


    reverse(nums, 0, nums.size() - 1);
    reverse(nums, 0, k - 1);
    reverse(nums, k, nums.size() - 1);

    return;
}

void reverse(vector<int>& nums, int start, int end) {
    while (start < end) {
        int temp = nums[start];
        nums[start] = nums[end];
        nums[end] = temp;

        start++;
        end--;
    }

    return;
}

File: LeetCode/Interview Questions/JS/RotateArray.js


let rotate = (array,k) => {
    k%=array.length;
    reverse(array, 0, array.length-1);
    reverse(array, 0, k-1);
    reverse(array, k, array.length-1);

};

let reverse = (array, low, high) => {
    while(low<high){
        let tmp = array[high];
        array[high] = array[low];
        array[low] = tmp;
        high--;
        low++;
    }
};



let array1 = [1,2,3,4,5,6];
rotate(array1, 3);
array1.forEach(element => {
    console.log(element);
});
